models/issue_watch.go

```diff
@@ -5,86 +5,133 @@
 package models
 
 import (
+	"fmt"
+
+	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/timeutil"
 )
 
+// IssueWatchMode specifies what kind of watch the user has on a issue
+type IssueWatchMode int8
+
+const (
+	// IssueWatchModeNone don't watch
+	IssueWatchModeNone IssueWatchMode = iota // 0
+	// IssueWatchModeNormal watch issue (from other sources)
+	IssueWatchModeNormal // 1
+	// IssueWatchModeDont explicit don't auto-watch
+	IssueWatchModeDont // 2
+	// IssueWatchModeAuto watch issue (from AutoWatchOnChanges)
+	IssueWatchModeAuto // 3
+)
+
 // IssueWatch is connection request for receiving issue notification.
 type IssueWatch struct {
 	ID          int64              `xorm:"pk autoincr"`
 	UserID      int64              `xorm:"UNIQUE(watch) NOT NULL"`
 	IssueID     int64              `xorm:"UNIQUE(watch) NOT NULL"`
-	IsWatching  bool               `xorm:"NOT NULL"`
+	Mode        IssueWatchMode     `xorm:"NOT NULL DEFAULT 1"`
 	CreatedUnix timeutil.TimeStamp `xorm:"created NOT NULL"`
 	UpdatedUnix timeutil.TimeStamp `xorm:"updated NOT NULL"`
 }
 
 // IssueWatchList contains IssueWatch
 type IssueWatchList []*IssueWatch
 
-// CreateOrUpdateIssueWatch set watching for a user and issue
-func CreateOrUpdateIssueWatch(userID, issueID int64, isWatching bool) error {
-	iw, exists, err := getIssueWatch(x, userID, issueID)
+// CreateOrUpdateIssueWatchMode set IssueWatchMode for a user and issue
+func CreateOrUpdateIssueWatchMode(userID, issueID int64, mode IssueWatchMode) error {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+	err := createOrUpdateIssueWatchMode(sess, userID, issueID, mode)
+	if err != nil {
+		return err
+	}
+	return sess.Commit()
+}
+
+func createOrUpdateIssueWatchMode(e Engine, userID, issueID int64, mode IssueWatchMode) error {
+	iw, exists, err := getIssueWatch(e, userID, issueID)
 	if err != nil {
 		return err
 	}
 
-	if !exists {
+	if !exists && mode != IssueWatchModeNone {
 		iw = &IssueWatch{
-			UserID:     userID,
-			IssueID:    issueID,
-			IsWatching: isWatching,
+			UserID:  userID,
+			IssueID: issueID,
+			Mode:    mode,
 		}
-
-		if _, err := x.Insert(iw); err != nil {
+		if _, err = e.Insert(iw); err != nil {
 			return err
 		}
 	} else {
-		iw.IsWatching = isWatching
-
-		if _, err := x.ID(iw.ID).Cols("is_watching", "updated_unix").Update(iw); err != nil {
+		if mode != IssueWatchModeNone {
+			iw.Mode = mode
+			if _, err = e.ID(iw.ID).Cols("updated_unix", "mode").Update(iw); err != nil {
+				return err
+			}
+		}
+		if _, err = e.ID(iw.ID).Delete(iw); err != nil {
```

Comment:
Did you mean `else if` here?

Reply:
no but I changed the function ... (IssueWatchModeNone is not here anymore)